sdk/monitor/monitor-opentelemetry/src/metrics/performanceCounters.ts

```diff
@@ -0,0 +1,392 @@
+      PerformanceCounterMetricNames.PROCESSOR_TIME,
+      {
+        description: "Processor time as a percentage",
+        valueType: ValueType.DOUBLE,
+      },
+    );
+    this.processTimeGauge = this.meter.createObservableGauge(
+      PerformanceCounterMetricNames.PROCESS_TIME,
+      {
+        description: "Process CPU usage as a percentage",
+        valueType: ValueType.DOUBLE,
+      },
+    );
+    this.processTimeNormalizedGauge = this.meter.createObservableGauge(
+      PerformanceCounterMetricNames.PROCESS_TIME_NORMALIZED,
+      {
+        description: "Process CPU usage normalized as a percentage",
+        valueType: ValueType.DOUBLE,
+      },
+    );
+    this.exceptionCountGauge = this.meter.createObservableGauge(
+      PerformanceCounterMetricNames.EXCEPTION_RATE,
+      {
+        description: "Exception Rate",
+        valueType: ValueType.DOUBLE,
+      },
+    );
+
+    // Add callbacks
+    this.requestRateGaugeCallback = this.getRequestRate.bind(this);
+    this.memoryPrivateBytesGaugeCallback = this.getPrivateMemory.bind(this);
+    this.memoryAvailableBytesGaugeCallback = this.getAvailableMemory.bind(this);
+    this.processorTimeGaugeCallback = this.getProcessorTime.bind(this);
+    this.processTimeGaugeCallback = this.getProcessTime.bind(this);
+    this.processTimeNormalizedGaugeCallback = this.getNormalizedProcessTime.bind(this);
+    this.exceptionCountGaugeCallback = this.getExceptionRate.bind(this);
+
+    this.memoryPrivateBytesGauge.addCallback(this.memoryPrivateBytesGaugeCallback);
+    this.memoryAvailableBytesGauge.addCallback(this.memoryAvailableBytesGaugeCallback);
+    this.processTimeGauge.addCallback(this.processTimeGaugeCallback);
+    this.processTimeNormalizedGauge.addCallback(this.processTimeNormalizedGaugeCallback);
+    this.processorTimeGauge.addCallback(this.processorTimeGaugeCallback);
+    this.requestRateGauge.addCallback(this.requestRateGaugeCallback);
+    this.exceptionCountGauge.addCallback(this.exceptionCountGaugeCallback);
+  }
+
+  /**
+   * Shutdown Meter Provider it will return no-op Meters after being called.
+   */
+  public async shutdown(): Promise<void> {
+    return this.meterProvider.shutdown();
+  }
+
+  /**
+   * Force flush Meter Provider.
+   */
+  public async flush(): Promise<void> {
+    return this.meterProvider.forceFlush();
+  }
+
+  /**
+   *Get OpenTelemetry MeterProvider
+   */
+  public getMeterProvider(): MeterProvider {
+    return this.meterProvider;
+  }
+
+  /**
+   * Record Span metrics
+   * @internal
+   */
+  public recordSpan(span: ReadableSpan): void {
+    if (span.kind !== SpanKind.SERVER) {
+      return;
+    }
+    this.totalCount++;
+    const durationMs = span.duration[0];
+    this.intervalExecutionTime += durationMs;
+    this.requestDurationHistogram.record(durationMs);
+    if (span.events) {
+      span.events.forEach((event: TimedEvent) => {
+        event.attributes = event.attributes || {};
```

Comment:
Won't this have a side effect of "setting" the `event.attributes` to an empty object. Is this expected?

Reply:
We do this since we'll likely set semantic convention values in the `event.attributes` in the exporter on these events. If we leave out this line, I believe attributes will remain undefined later and we'll miss setting those values.